{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BioE 131 LAB 4 | Connor Tou"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As shown in the below table, 4 enzymes were chosen (ordered) from 3 pathways by looking at the graphical pathways on KEGG. The EC (Enzyme Comission) Number is shown. \n",
    "\n",
    "| PATHWAY           | ENZYME 1 | ENZYME 2 | ENZYME 3 | ENZYME 4 |\n",
    "|-------------------|----------|----------|----------|----------|\n",
    "| Glycolysis        | 5.4.2.2  | 5.3.1.9  | 4.1.1.49 | 6.2.1.13 |\n",
    "| TCA               | 1.1.5.4  | 4.1.1.31 | 4.1.1.49 | 4.1.1.32 |\n",
    "| Pentose Phosphate | 2.7.1.13 | 4.1.2.43 | 5.3.1.6  | 5.4.2.7  |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DATABASE SCHEMA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GENES TABLE\n",
    "\n",
    "| ID | NAME | ORGNANISM |   |   |\n",
    "|----|------|-----------|---|---|\n",
    "|    |      |           |   |   |\n",
    "|    |      |           |   |   |\n",
    "|    |      |           |   |   |\n",
    "\n",
    "36 genes will be stored in this table: 3 pathways, 4 genes, 3 organisms (homo sapiens, drosophilia melanogaster, and escherichia coli K-12 MG165)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PATHWAYS TABLE\n",
    "\n",
    "| ID | NAME | DESCRIPTION |\n",
    "|----|------|-------------|\n",
    "|    |      |             |\n",
    "|    |      |             |\n",
    "\n",
    "Information for the 3 pathways are stored here. Note that Descriptions were taken directly from KEGG. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ENZYME TABLE\n",
    "\n",
    "| ID | NAME | DESCRIPTION | Enzyme Commission # |\n",
    "|----|------|-------------|---------------------|\n",
    "|    |      |             |                     |\n",
    "|    |      |             |                     |\n",
    "\n",
    "The information for 12 enzymes will be stored here. EC # is used. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ASSOCIATIVE TABLE 1: PATHWAY_ENZYME TABLE\n",
    "\n",
    "| ID | PATHWAY | ENZYME | ORDER |\n",
    "|----|---------|--------|-------|\n",
    "|    |         |        |       |\n",
    "|    |         |        |       |\n",
    "\n",
    "The relationship between pathways and enzymes (and the enzyme's order in a given pathway) are stored here. **Note that some enzymes belong to many pathways - this is a one to many relationship.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ASSOCIATIVE TABLE 2: ENZYME_GENE TABLE\n",
    "\n",
    "| ID | ENZYME COMMISSION # | ORGANISM | GENE |\n",
    "|----|---------------------|----------|------|\n",
    "|    |                     |          |      |\n",
    "|    |                     |          |      |\n",
    "\n",
    "The relationship between enzymes and genes are stored here. **Note that a single gene can be part of many enzymes (through alternativ and a single enzyme contains many coding genes. This is a many to many relationship.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DROPPING TABLES...BEEP, BOOP,BEEP, BOOP...MAKING NEW ONES...BEEP, BOOP, BEEP, BOOP\n",
      "Pathway: Glycolysis; EC: 5.4.2.2; Organism: homo sapiens\n",
      "Pathway: Glycolysis; EC: 5.4.2.2; Organism: escherichia coli\n",
      "Pathway: Glycolysis; EC: 5.4.2.2; Organism: drosophila\n",
      "Pathway: Glycolysis; EC: 5.3.1.9; Organism: homo sapiens\n",
      "Pathway: Glycolysis; EC: 5.3.1.9; Organism: escherichia coli\n",
      "Pathway: Glycolysis; EC: 5.3.1.9; Organism: drosophila\n",
      "Pathway: Glycolysis; EC: 4.1.1.49; Organism: homo sapiens\n",
      "Pathway: Glycolysis; EC: 4.1.1.49; Organism: escherichia coli\n",
      "Pathway: Glycolysis; EC: 4.1.1.49; Organism: drosophila\n",
      "Pathway: Glycolysis; EC: 6.2.1.13; Organism: homo sapiens\n",
      "Pathway: Glycolysis; EC: 6.2.1.13; Organism: escherichia coli\n",
      "Pathway: Glycolysis; EC: 6.2.1.13; Organism: drosophila\n",
      "Pathway: TCA cycle; EC: 1.1.5.4; Organism: homo sapiens\n",
      "Pathway: TCA cycle; EC: 1.1.5.4; Organism: escherichia coli\n",
      "Pathway: TCA cycle; EC: 1.1.5.4; Organism: drosophila\n",
      "Pathway: TCA cycle; EC: 4.1.1.31; Organism: homo sapiens\n",
      "Pathway: TCA cycle; EC: 4.1.1.31; Organism: escherichia coli\n",
      "Pathway: TCA cycle; EC: 4.1.1.31; Organism: drosophila\n",
      "Pathway: TCA cycle; EC: 4.1.1.49; Organism: homo sapiens\n",
      "Pathway: TCA cycle; EC: 4.1.1.49; Organism: escherichia coli\n",
      "Pathway: TCA cycle; EC: 4.1.1.49; Organism: drosophila\n",
      "Pathway: TCA cycle; EC: 4.1.1.32; Organism: homo sapiens\n",
      "Pathway: TCA cycle; EC: 4.1.1.32; Organism: escherichia coli\n",
      "Pathway: TCA cycle; EC: 4.1.1.32; Organism: drosophila\n",
      "Pathway: Pentose Phosphate; EC: 2.7.1.13; Organism: homo sapiens\n",
      "Pathway: Pentose Phosphate; EC: 2.7.1.13; Organism: escherichia coli\n",
      "Pathway: Pentose Phosphate; EC: 2.7.1.13; Organism: drosophila\n",
      "Pathway: Pentose Phosphate; EC: 4.1.2.43; Organism: homo sapiens\n",
      "Pathway: Pentose Phosphate; EC: 4.1.2.43; Organism: escherichia coli\n",
      "Pathway: Pentose Phosphate; EC: 4.1.2.43; Organism: drosophila\n",
      "Pathway: Pentose Phosphate; EC: 5.3.1.6; Organism: homo sapiens\n",
      "Pathway: Pentose Phosphate; EC: 5.3.1.6; Organism: escherichia coli\n",
      "Pathway: Pentose Phosphate; EC: 5.3.1.6; Organism: drosophila\n",
      "Pathway: Pentose Phosphate; EC: 5.4.2.7; Organism: homo sapiens\n",
      "Pathway: Pentose Phosphate; EC: 5.4.2.7; Organism: escherichia coli\n",
      "Pathway: Pentose Phosphate; EC: 5.4.2.7; Organism: drosophila\n",
      "\n",
      "Pathways table:\n",
      "   id               name                                        description\n",
      "0   1         Glycolysis  Glycolysis is the process of converting glucos...\n",
      "1   2          TCA cycle  The citrate cycle (TCA cycle, Krebs cycle) is ...\n",
      "2   3  Pentose Phosphate  The pentose phosphate pathway is a process of ...\n",
      "\n",
      "Genes table:\n",
      "    id        name          organism  \\\n",
      "0    1    X68132.1      homo sapiens   \n",
      "1    2    L39517.1  escherichia coli   \n",
      "2    3    X64011.1        drosophila   \n",
      "3    4    Z21512.1      homo sapiens   \n",
      "4    5      T44203  escherichia coli   \n",
      "5    6  AW258048.1        drosophila   \n",
      "6    7  CQ421109.1      homo sapiens   \n",
      "7    8    R86926.1  escherichia coli   \n",
      "8    9  CF389836.1        drosophila   \n",
      "9   10  BH517191.1      homo sapiens   \n",
      "10  11  AF059503.1  escherichia coli   \n",
      "11  12  BU660270.1      homo sapiens   \n",
      "12  13    R87889.1  escherichia coli   \n",
      "13  14  CF388434.1        drosophila   \n",
      "14  15  BM215258.2      homo sapiens   \n",
      "15  16      T44113  escherichia coli   \n",
      "16  17  CF389472.1        drosophila   \n",
      "17  18  CQ421109.1      homo sapiens   \n",
      "18  19    R86926.1  escherichia coli   \n",
      "19  20  CF389836.1        drosophila   \n",
      "20  21    X68851.1      homo sapiens   \n",
      "21  22  AL135793.1        drosophila   \n",
      "22  23  AJ410030.1      homo sapiens   \n",
      "23  24  BX507314.1      homo sapiens   \n",
      "24  25    X64244.1      homo sapiens   \n",
      "25  26      T44279  escherichia coli   \n",
      "26  27  AJ269903.1        drosophila   \n",
      "27  28    X60334.1      homo sapiens   \n",
      "28  29      T44637  escherichia coli   \n",
      "29  30  CF388555.1        drosophila   \n",
      "\n",
      "                                             sequence  \n",
      "0   TTCCCTGCTCGTATGTCTTGAGATTGGGCCTTAAATAATATGCAGC...  \n",
      "1   ACTTACTACTGTGCCCTCTCTCTTTACTGGGGGATACAGGCACCGA...  \n",
      "2   CGTTATTTAAGGTGTTACATAGTTCTATGGAAATAGGGTCTATACC...  \n",
      "3   ATCGATTTGAGAGGTGATAGGTGATTAGCGGGGGGAGATGAAAAGT...  \n",
      "4   GAAAAACAGACAGCGGAGGACTAAAACAACTAGCCACAACACAACG...  \n",
      "5   TGGCGGAACCAAGTACAACATTTGAAAAGGACCTGTCAACTGTAGA...  \n",
      "6   TAGGGCTTTTTGGAGCTCCCCGCGGTGGCGGCCGCCCGGGCAGGTN...  \n",
      "7   ATNAAAATAAAGGATATTTTTTTTATTCAAGTAACTGCAAATAGGA...  \n",
      "8   TTGTGACTGAGCTTAAATCGAATGATATAATCGATAATGGCGCCTA...  \n",
      "9   GTCATAGATAGATCACCCAGGTTCGGGTCCATAAGCAGTGACAATT...  \n",
      "10  CCTGGCGGCGGCGGTGCAGCGACAGACCGAGGCGACCCCGTCAACG...  \n",
      "11  AATCTACTAACATTCTATTAGCAATCATGCTCCGAGGTTTGATTTT...  \n",
      "12  AAACGACCAGAATCCAGGTTCCCAGTTCAGCCAGGCCCTGGCATTC...  \n",
      "13  GTACATTTTTGCAATGATAGGATTGCAAGTAGACTCCAGCTCTTTC...  \n",
      "14  GGTTTTATTCCTACAGAAAGTTGATTTTTATGACTGTTTCTAATTT...  \n",
      "15  CCTGANTCGTCTTCTTCGATAATCTCCCTCTTGTTTCAATCAAGAA...  \n",
      "16  AGTGGGCATGCAGGCTTTGTGATGAATTGGTTAAGTGGGCATGCAG...  \n",
      "17  TAGGGCTTTTTGGAGCTCCCCGCGGTGGCGGCCGCCCGGGCAGGTN...  \n",
      "18  ATNAAAATAAAGGATATTTTTTTTATTCAAGTAACTGCAAATAGGA...  \n",
      "19  TTGTGACTGAGCTTAAATCGAATGATATAATCGATAATGGCGCCTA...  \n",
      "20  ATCTGTACGATCCGTCGTGATTGTGTTACAAACCAAAAATCGCTTT...  \n",
      "21  CCTGGTTGGAAGCCACCAAGAAGGCCCCCAGAAGGCAACAGGTGGG...  \n",
      "22  CCAAAAAGCTTATTCGCCCCCCAACTATTTATCCATTCTATCCCCC...  \n",
      "23  AGGGTCACCAGGAATTTTCTGCGGTTTGCATTTTTCCCATAAGAAG...  \n",
      "24  ATGATGGTGCTGGACAAGGAGGACGGCGTGCCGATGCTCTCCGTCC...  \n",
      "25  GTAGCAGCTGGTGGTTCCATGGGGAGCAATCACAACTGNTTGCAGC...  \n",
      "26  GGGGGTCCTAAATTAAGGGGTAATGTTACAAGTAATATAAAGTTCC...  \n",
      "27  GAGGTCCAGCTGCAACAGTCTGGACCTGAACTGGTAGAGCCTGGGA...  \n",
      "28  CACAAACAAACAATAACTAAAATGGCTTTGTCTTCCAAAACCCTAA...  \n",
      "29  TATGTTTCAAGCTCTAATTCCAACTAGGAGATTCAAGTTTGATGTT...  \n",
      "\n",
      "Enzymes table:\n",
      "         ec      name                                        description\n",
      "0   5.4.2.2  PVH54613              phosphoglucomutase [Escherichia coli]\n",
      "1   5.3.1.9  QCJ61871   glucose-6-phosphate isomerase [Escherichia coli]\n",
      "2   1.1.5.4  QCJ59454  malate dehydrogenase (quinone) [Escherichia coli]\n",
      "3  4.1.1.31  AAA24414  phosphoenolpyruvate carboxylase mutant (EC 4.1...\n",
      "4  4.1.1.49  QCJ58280  phosphoenolpyruvate carboxykinase (ATP) [Esche...\n",
      "5  4.1.1.32    1II2_B         Chain B, Phosphoenolpyruvate Carboxykinase\n",
      "6   5.3.1.6  PVH52263    ribose-5-phosphate isomerase [Escherichia coli]\n",
      "7   5.4.2.7  QCJ61557              phosphopentomutase [Escherichia coli]\n",
      "\n",
      "Pathway enzyme associated table:\n",
      "    id            pathway    enzyme  pathway_order\n",
      "0    1         Glycolysis   5.4.2.2              1\n",
      "1    2         Glycolysis   5.3.1.9              2\n",
      "2    3         Glycolysis  4.1.1.49              3\n",
      "3    4         Glycolysis  6.2.1.13              4\n",
      "4    5          TCA cycle   1.1.5.4              1\n",
      "5    6          TCA cycle  4.1.1.31              2\n",
      "6    7          TCA cycle  4.1.1.49              3\n",
      "7    8          TCA cycle  4.1.1.32              4\n",
      "8    9  Pentose Phosphate  2.7.1.13              1\n",
      "9   10  Pentose Phosphate  4.1.2.43              2\n",
      "10  11  Pentose Phosphate   5.3.1.6              3\n",
      "11  12  Pentose Phosphate   5.4.2.7              4\n",
      "\n",
      "Enzyme genes associated table:\n",
      "    id        ec          organism        gene\n",
      "0    1   5.4.2.2      homo sapiens    X68132.1\n",
      "1    2   5.4.2.2  escherichia coli    L39517.1\n",
      "2    3   5.4.2.2        drosophila    X64011.1\n",
      "3    4   5.3.1.9      homo sapiens    Z21512.1\n",
      "4    5   5.3.1.9  escherichia coli      T44203\n",
      "5    6   5.3.1.9        drosophila  AW258048.1\n",
      "6    7  4.1.1.49      homo sapiens  CQ421109.1\n",
      "7    8  4.1.1.49  escherichia coli    R86926.1\n",
      "8    9  4.1.1.49        drosophila  CF389836.1\n",
      "9   10  6.2.1.13      homo sapiens  BH517191.1\n",
      "10  11  6.2.1.13  escherichia coli  AF059503.1\n",
      "11  12   1.1.5.4      homo sapiens  BU660270.1\n",
      "12  13   1.1.5.4  escherichia coli    R87889.1\n",
      "13  14   1.1.5.4        drosophila  CF388434.1\n",
      "14  15  4.1.1.31      homo sapiens  BM215258.2\n",
      "15  16  4.1.1.31  escherichia coli      T44113\n",
      "16  17  4.1.1.31        drosophila  CF389472.1\n",
      "17  18  4.1.1.49      homo sapiens  CQ421109.1\n",
      "18  19  4.1.1.49  escherichia coli    R86926.1\n",
      "19  20  4.1.1.49        drosophila  CF389836.1\n",
      "20  21  4.1.1.32      homo sapiens    X68851.1\n",
      "21  22  4.1.1.32        drosophila  AL135793.1\n",
      "22  23  2.7.1.13      homo sapiens  AJ410030.1\n",
      "23  24  4.1.2.43      homo sapiens  BX507314.1\n",
      "24  25   5.3.1.6      homo sapiens    X64244.1\n",
      "25  26   5.3.1.6  escherichia coli      T44279\n",
      "26  27   5.3.1.6        drosophila  AJ269903.1\n",
      "27  28   5.4.2.7      homo sapiens    X60334.1\n",
      "28  29   5.4.2.7  escherichia coli      T44637\n",
      "29  30   5.4.2.7        drosophila  CF388555.1\n"
     ]
    }
   ],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import sys\n",
    "\n",
    "Entrez.email = \"connortou@berkeley.edu\"\n",
    "conn = sqlite3.connect('Lab4.db')\n",
    "c = conn.cursor()\n",
    "\n",
    "# If these SQL tables already exist, delete them\n",
    "try:\n",
    "    print(\"DROPPING TABLES...BEEP..BOOP...MAKING NEW ONES...BEEP..BOOP..BEEP..BOOP\")\n",
    "    c.execute(\"DROP TABLE genes\")\n",
    "    c.execute(\"DROP TABLE pathways\")\n",
    "    c.execute(\"DROP TABLE enzymes\")\n",
    "    c.execute(\"DROP TABLE pathway_enzyme\")\n",
    "    c.execute(\"DROP TABLE enzyme_gene\")\n",
    "except:\n",
    "    print(\"MAKING TABLES...BEEP...BOOP\")\n",
    "    \n",
    "    \n",
    "# Create these SQL tables\n",
    "c.execute(\"\"\"CREATE TABLE genes (id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                                name TEXT,\n",
    "                                organism TEXT,\n",
    "                                sequence TEXT);\"\"\")\n",
    "\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE pathways (id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                                    name TEXT,\n",
    "                                    description TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzymes (ec TEXT PRIMARY KEY,\n",
    "                                    name TEXT,\n",
    "                                    description TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE pathway_enzyme (id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                                            pathway TEXT,\n",
    "                                            enzyme TEXT,\n",
    "                                            pathway_order INTEGER);\"\"\")\n",
    "\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzyme_gene (id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                                            ec TEXT,\n",
    "                                            organism TEXT,\n",
    "                                            gene TEXT);\"\"\")\n",
    "\n",
    "# Considered organisms, pathways and enzymes. Enzymes are given in order of occurence in the pathway.\n",
    "organism = ['homo sapiens', 'escherichia coli', 'drosophila']\n",
    "\n",
    "path = {'Glycolysis': ['5.4.2.2','5.3.1.9', '4.1.1.49','6.2.1.13'], \n",
    "      'TCA cycle':['1.1.5.4', '4.1.1.31','4.1.1.49','4.1.1.32'],\n",
    "      'Pentose Phosphate':['2.7.1.13', '4.1.2.43', '5.3.1.6', '5.4.2.7']}\n",
    "\n",
    "# Enter Descriptions (Pulled from KEGG)\n",
    "c.execute(\"\"\"INSERT INTO pathways (name,\n",
    "                                    description)\n",
    "                            VALUES\n",
    "                                    ('Glycolysis',\n",
    "                                    'Glycolysis is the process of converting glucose into pyruvate and generating small \n",
    "                                    amounts of ATP (energy) and NADH (reducing power). It is a central pathway that \n",
    "                                    produces important precursor metabolites: six-carbon compounds of glucose-6P and \n",
    "                                    fructose-6P and three-carbon compounds of glycerone-P, glyceraldehyde-3P, \n",
    "                                    glycerate-3P, phosphoenolpyruvate, and pyruvate. Acetyl-CoA, another important \n",
    "                                    precursor metabolite, is produced by oxidative decarboxylation of pyruvate. When the\n",
    "                                    enzyme genes of this pathway are examined in completely sequenced genomes, the \n",
    "                                    reaction steps of three-carbon compounds from glycerone-P to pyruvate form a \n",
    "                                    conserved core module, which is found in almost all organisms and which sometimes\n",
    "                                    contains operon structures in bacterial genomes. Gluconeogenesis is a synthesis \n",
    "                                    pathway of glucose from noncarbohydrate precursors. It is essentially a reversal of \n",
    "                                    glycolysis with minor variations of alternative paths.')\n",
    "                                    ;\"\"\")\n",
    "\n",
    "c.execute(\"\"\"INSERT INTO pathways (name,\n",
    "                                    description)\n",
    "                            VALUES\n",
    "                                    ('TCA cycle',\n",
    "                                    'The citrate cycle (TCA cycle, Krebs cycle) is an important aerobic pathway for the \n",
    "                                    final steps of the oxidation of carbohydrates and fatty acids. The cycle starts with \n",
    "                                    acetyl-CoA, the activated form of acetate, derived from glycolysis and pyruvate \n",
    "                                    oxidation for carbohydrates and from beta oxidation of fatty acids. The two-carbon \n",
    "                                    acetyl group in acetyl-CoA is transferred to the four-carbon compound of oxaloacetate \n",
    "                                    to form the six-carbon compound of citrate. In a series of reactions two carbons in \n",
    "                                    citrate are oxidized to CO2 and the reaction pathway supplies NADH for use in the \n",
    "                                    oxidative phosphorylation and other metabolic processes. The pathway also supplies \n",
    "                                    important precursor metabolites including 2-oxoglutarate. At the end of the cycle the \n",
    "                                    remaining four-carbon part is transformed back to oxaloacetate. According to the \n",
    "                                    genome sequence data, many organisms seem to lack genes for the full cycle, but \n",
    "                                    contain genes for specific segments.')\n",
    "                                    ;\"\"\")\n",
    "\n",
    "c.execute(\"\"\"INSERT INTO pathways (name,\n",
    "                                    description)\n",
    "                            VALUES\n",
    "                                    ('Pentose Phosphate',\n",
    "                                    'The pentose phosphate pathway is a process of glucose turnover that produces NADPH \n",
    "                                    as reducing equivalents and pentoses as essential parts of nucleotides. There are two \n",
    "                                    different phases in the pathway. One is irreversible oxidative phase in which \n",
    "                                    glucose-6P is converted to ribulose-5P by oxidative decarboxylation, and NADPH is \n",
    "                                    generated. The other is reversible non-oxidative phase in which phosphorylated sugars \n",
    "                                    are interconverted to generate xylulose-5P, ribulose-5P, and ribose-5P. \n",
    "                                    Phosphoribosyl pyrophosphate (PRPP) formed from ribose-5P is an activated compound \n",
    "                                    used in the biosynthesis of histidine and purine/pyrimidine nucleotides. This pathway \n",
    "                                    map also shows the Entner-Doudoroff pathway where 6-P-gluconate is dehydrated and then\n",
    "                                    cleaved into pyruvate and glyceraldehyde-3P.')\n",
    "                                    ;\"\"\")\n",
    "\n",
    "# Extract data from Entrez automatically to fill other tables\n",
    "for (pathway, enzyme) in path.items():\n",
    "    # number of pathway enzyme\n",
    "    num_enzyme = 1\n",
    "    for e in enzyme:\n",
    "        # Create associated pathway linking pathways and enzymes\n",
    "        c.execute(\"\"\"INSERT INTO pathway_enzyme (pathway,\n",
    "                                                enzyme,\n",
    "                                                pathway_order)\n",
    "                                        VALUES\n",
    "                                                (?,\n",
    "                                                ?,\n",
    "                                                ?);\"\"\" ,\n",
    "                                                (pathway,e,num_enzyme))\n",
    "        num_enzyme +=1\n",
    "             \n",
    "       \n",
    "        \n",
    "        # orgnanism sequence\n",
    "        for o in organism:\n",
    "            # The current search is printed to give the user an indication of progress\n",
    "            print('Pathway: ' + pathway + '; EC: ' + e + '; Organism: ' + o)\n",
    "\n",
    "             # Search for info using Entrez\n",
    "            handle = Entrez.esearch(db = 'protein',\n",
    "                                    term = 'EC: ' + e + ' AND ' + o,\n",
    "                                    sort = 'relevance',\n",
    "                                    idtype = 'acc',\n",
    "                                    retmax = 1)\n",
    "            \n",
    "            for i in Entrez.read(handle)['IdList']:\n",
    "                handle = Entrez.efetch(db = 'protein', id = i, rettype = 'gb', retmode = 'text')\n",
    "                record = SeqIO.read(handle, \"genbank\")\n",
    "                handle.close()\n",
    "                c.execute(\"\"\"INSERT OR REPLACE INTO enzymes (name,\n",
    "                                                description,\n",
    "                                                ec)\n",
    "                                        VALUES\n",
    "                                                (?,\n",
    "                                                ?,\n",
    "                                                ?);\"\"\" ,\n",
    "                                                (record.name,record.description,e))\n",
    "                \n",
    "            # The corresponding nucleotide sequence information is searched for using Entrez\n",
    "            # The sequence is found using the gene database, the rest of the information is found using the nucleotide one\n",
    "            handle = Entrez.esearch(db = 'gene',\n",
    "                                    term = e + ' ' + o,\n",
    "                                    sort = 'relevance',\n",
    "                                    idtype = 'acc',\n",
    "                                    retmax = 1)\n",
    "            \n",
    "            \n",
    "            for i in Entrez.read(handle)['IdList']:\n",
    "                handle = Entrez.efetch(db = 'nucleotide', id = i, rettype = 'gb', retmode = 'text')\n",
    "                record = SeqIO.read(handle, \"genbank\")\n",
    "                handle.close()\n",
    "\n",
    "                c.execute(\"\"\"INSERT INTO genes (name,\n",
    "                                                organism,\n",
    "                                                sequence)\n",
    "                                        VALUES\n",
    "                                                (?,\n",
    "                                                ?,\n",
    "                                                ?);\"\"\" ,\n",
    "                                                (record.id, o, str(record.seq)))\n",
    "        \n",
    "                c.execute(\"\"\"INSERT INTO enzyme_gene (ec,\n",
    "                                                    organism,\n",
    "                                                    gene)\n",
    "                                            VALUES\n",
    "                                                    (?,\n",
    "                                                    ?,\n",
    "                                                    ?);\"\"\" ,\n",
    "                                                    (e,o,record.id))\n",
    "            \n",
    "\n",
    "# Update tables\n",
    "conn.commit()\n",
    "\n",
    "# Print tables\n",
    "print(\"\\nPathways table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM pathways\", conn))\n",
    "print(\"\\nGenes table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM genes\", conn))\n",
    "print(\"\\nEnzymes table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM enzymes\", conn))\n",
    "print(\"\\nPathway enzyme associated table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM pathway_enzyme\", conn))\n",
    "print(\"\\nEnzyme genes associated table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM enzyme_gene\", conn))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Try UCSC, Entrez, KEGG, and Reactome. Notice how each database presents the information in a different way. \n",
    "\n",
    "1.) Do you have any preference? \n",
    "* I like the interactive nature of Reactome; however, I find it a bit clunkier and harder to use/navigate. For utility purposes, I prefer KEGG or UCSC.\n",
    "\n",
    "2.) Which seems to be the most complete? \n",
    "* KEGG and UCSC both have extensive data and seems the most complete. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# END OF LAB 4"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
